### PR TITLE
[Xrd] Wait for the current PollE poll to finish after removing a Link

### PR DESCRIPTION
**Pull request by smithdh**
This is a change to avoid issue #1928.

### CODE REVIEW

File: src/Xrd/XrdPollE.icc [hunk 1]
```diff
@@ -79,9 +88,30 @@ XrdPoll *XrdPoll::newPoller(int pollid, int maxfd)
 XrdPollE::~XrdPollE()
 {
    if (PollTab) free(PollTab);
+   if (WaitFd >= 0) close(WaitFd);
    if (PollDfd >= 0) close(PollDfd);
 }
-  
+
+/******************************************************************************/
+/*                              A d d E v e n t F d                           */
```
Comment (abh3) [inline src/Xrd/XrdPollE.icc:96]: This should be AddWaitFD (looks like you must have changed your mind at some point).

File: src/Xrd/XrdPollE.icc [hunk 2]
```diff
@@ -164,6 +194,44 @@ void XrdPollE::Exclude(XrdPollInfo &pInfo)
...
+// Wait to make sure the poll thread has completed handling the last set of
+// events which may have included this Link. The handing includes examining the
+// Link's PollInfo before deciding if to schedule the Link. After we return,
+// the PollInfo may be reset and the Link could be subsequently reused.
+//
+   WaitPoller();
+}
+
+/******************************************************************************/
+/*                          H a n d l e E v e n t F d                         */
```
Comment (abh3) [inline src/Xrd/XrdPollE.icc:207]: This should be HandleWaitFD

File: src/Xrd/XrdPollE.icc [hunk 3]
```diff
@@ -272,9 +351,28 @@ void XrdPollE::Start(XrdSysSemaphore *syncsem, int &retcode)
        //
        if (num2sched == 1) Sched.Schedule(jfirst);
           else if (num2sched) Sched.Schedule(num2sched, jfirst, jlast);
+
+       if (haveWaiters) HandleWaitFd(waitFdEvents);
       } while(1);
 }
 
+/******************************************************************************/
+/*                             W a i t E v e n t F d                          */
```
Comment (abh3) [inline src/Xrd/XrdPollE.icc:360]: This should ne WaitPoller (perhaps a better name would be Wait4Poller)

**Review by abh3 — CHANGES_REQUESTED** [review]
Just comment changes are needed. Presumably the code was tested with the "sleeps" that caused it to fail and now it doesn't fail, right? The only other suggestion is consider repackaging the "eventfd' processing into a separate class so that it can be used elsewhere should the need arise.

**Comment by smithdh** [comment]
Thanks very much for the comments. I've changed those; the banners and renamed Wait4Poller. I've tried it with my "sleeps" it solves what I could provoke. (I don't know how to reproduce the original production-system problem, as that seems to be quite rare).

Repackaging as a separate class is likely a better way to make it. However I'm not sure I can find a nice general interface for it. (It would probably be used with an epoll-poller, as eventfd is linux specific like epoll). I don't think we have the same problem in XrdSysIOEventsPollE because of how that poller works. So I propose to leave it here (not separated into another class) now.

**Pull request merged**